demo/runtime/src/runtime/council.rs

```diff
@@ -0,0 +1,1550 @@
+	/// invoke `finalise_vote`.
+	pub fn set_term_duration(count: BlockNumber) {
+		storage::put(TERM_DURATION, &count);
+	}
+}
+
+pub mod internal {
+	use super::*;
+	use demo_primitives::Proposal;
+	use dispatch::enact_proposal;
+
+	/// Check there's nothing to do this block
+	pub fn end_block() {
+		let block_number = system::block_number();
+		if block_number % voting_period() == 0 {
+			if let Some(number) = next_tally() {
+				if block_number == number {
+					start_tally();
+				}
+			}
+		}
+		if let Some((number, _, _)) = next_finalise() {
+			if block_number == number {
+				finalise_tally();
+			}
+		}
+	}
+}
+
+/// Remove a voter from the system. Trusts that voters()[index] != voter.
+fn remove_voter(voter: &AccountId, index: usize, mut voters: Vec<AccountId>) {
+	storage::put(VOTERS, &{ voters.swap_remove(index); voters });
+	storage::kill(&voter.to_keyed_vec(APPROVALS_OF));
+	storage::kill(&voter.to_keyed_vec(LAST_ACTIVE_OF));
+}
+
+/// Close the voting, snapshot the staking and the number of seats that are actually up for grabs.
+fn start_tally() {
+	let active_council = active_council();
+	let desired_seats = desired_seats() as usize;
+	let number = system::block_number();
+	let expiring = active_council.iter().take_while(|i| i.1 == number).cloned().collect::<Vec<_>>();
+	if active_council.len() - expiring.len() < desired_seats {
+		let empty_seats = desired_seats - (active_council.len() - expiring.len());
+		storage::put(NEXT_FINALISE, &(number + presentation_duration(), empty_seats as u32, expiring));
+
+		let voters: Vec<AccountId> = storage::get_or_default(VOTERS);
+		let votes: Vec<Balance> = voters.iter().map(staking::balance).collect();
+		storage::put(SNAPSHOTED_STAKES, &votes);
+
+		// initialise leaderboard.
+		let leaderboard_size = empty_seats + carry_count() as usize;
+		storage::put(LEADERBOARD, &vec![(0 as Balance, AccountId::default()); leaderboard_size]);
+	}
+}
+
+/// Finalise the vote, removing each of the `removals` and inserting `seats` of the most approved
+/// candidates in their place. If the total council members is less than the desired membership
+/// a new vote is started.
+/// Clears all presented candidates, returning the bond of the elected ones.
+fn finalise_tally() {
+	storage::kill(SNAPSHOTED_STAKES);
+	let (_, coming, expiring): (BlockNumber, u32, Vec<AccountId>) = storage::take(NEXT_FINALISE)
+		.expect("finalise can only be called after a tally is started.");
+	let leaderboard: Vec<(Balance, AccountId)> = storage::take(LEADERBOARD).unwrap_or_default();
+	let new_expiry = system::block_number() + term_duration();
+
+	// return bond to winners.
+	let candidacy_bond = candidacy_bond();
+	for &(_, ref w) in leaderboard.iter()
+		.rev()
+		.take_while(|&&(b, _)| b != 0)
+		.take(coming as usize)
+	{
+		staking::internal::set_balance(w, staking::balance(w) + candidacy_bond);
+	}
+
+	// set the new council.
+	let new_council: Vec<_> = active_council()
+		.into_iter()
+		.skip(expiring.len())
```

Comment:
shouldn't it rather skip any authorities who are expiring? where is it guaranteed that all expiring seats are at the front?

Reply:
`active_council` is kept ordered by expiry `BlockNumber`

Reply:
(there's a test to ensure that no referendum can be injected that is out of order.)

Reply:
then having the `expiring` value contain a vector of expiring account IDs is a red herring and should be replaced with an unsigned integer.

Reply:
it's now enforced. before the explicit sort it was possible to have an out-of-order council only by maliciously shortening the term duration.